### PR TITLE
Common test interface

### PR DESCRIPTION
**Pull request by benrudolph**
@dannyroberts Ok here's the initial version of the tests refactor. Not super in love with this, but I suppose it's a start. Went back and forth on how to do things so the commits got really messy. Right now the first commit is files that I didn't touch at all, so it's probably best to just look at the second commit. Not all tests are passing, had some trouble getting the assertDict to work properly, will be looking into that.

buddy: @czue

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 10 errors:

* corehq/form_processor/interfaces.py, line 76 - E127 continuation line over-indented for visual indent
* corehq/form_processor/tests/test_archive.py, line 3 - F401 'XFormArchived' imported but unused
* corehq/form_processor/tests/test_archive.py, line 3 - F401 'XFormInstance' imported but unused
* corehq/form_processor/tests/test_archive.py, line 5 - F401 'fetch_and_wrap_form' imported but unused
* corehq/form_processor/tests/test_edits.py, line 8 - F401 'CommCareCase' imported but unused
* corehq/form_processor/tests/test_edits.py, line 15 - F401 'GenericXFormInstance' imported but unused
* corehq/form_processor/tests/test_errors.py, line 3 - F401 'XFormError' imported but unused
* corehq/form_processor/tests/test_meta.py, line 24 - E231 missing whitespace after ','
* corehq/form_processor/tests/test_meta.py, line 25 - E231 missing whitespace after ','
* corehq/form_processor/tests/test_namespaces.py, line 3 - F401 'XFormInstance' imported but unused

File: corehq/form_processor/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    @staticmethod
+    def update_properties(xform, **properties):
+        for prop, value in properties.iteritems():
+            setattr(xform, prop, value)
+        xform.save()
+
+    @staticmethod
+    def post_xform(instance_xml, attachments=None, process=None,
+                            domain='test-domain'):
```
Comment (dimagimon) [inline corehq/form_processor/interfaces.py]: E127 continuation line over-indented for visual indent

**Comment by dannyroberts** [comment]
Feels a bit hard to review without a diff with the original tests. Any suggestions? I'm happy to sit with you if that's easier

**Comment by benrudolph** [comment]
Yea I was thinking about that. Maybe synchronous review could help

File: corehq/form_processor/generic.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
+import datetime
+from dimagi.ext.jsonobject import (
```
Comment (dimagimon) [inline corehq/form_processor/generic.py]: F401 'ObjectProperty' imported but unused

**Comment by benrudolph** [comment]
@dannyroberts this passes tests and i'm almost a fan of merging this and then iterating instead of getting this bit perfect. let me know if you think there are some blockers

File: corehq/form_processor/interfaces.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    @staticmethod
+    def get_xml_element(xform_generic):
+        xform = FormProcessorInterface._get_xform(xform_generic.id)
+        return xform.get_xml_element()
+
+    @staticmethod
+    @to_generic
+    def get_xform(xform_id):
+        return FormProcessorInterface._get_xform(xform_id)
```
Comment (dannyroberts) [inline corehq/form_processor/interfaces.py]: This is a nitpick, but I think it makes more sense to have this be a `@classmethod` and reference `FormProcessorInterface` as `cls`. (There's the conventions argument, and then there's the "what happens when you subclass this" argument, that may actually end up being relevant in this case, or maybe not, depending on how exactly we implement the two backends.)

Comment (benrudolph) [inline corehq/form_processor/interfaces.py]: good point, will update

**Comment by dannyroberts** [comment]
Looks solid to me. @snopoke @czue might be worth getting some more eyes on this since it'll kind of set the path forward for a few things that affect everyone.

@benrudolph correct me if I'm wrong, these are the basic principals used in this PR:

- All tests only access xforms through the `FormProcessorInterface` class methods.
- `FormProcessorInterface` methods only return `GenericXFormInstance`s etc. so tests never even get to put their grimy hands on real db models

This makes `FormProcessorInterface` kind of document everything we currently use. There's a number of ways we can go from here, but this feels like kind of the first step for all of them.

**Comment by snopoke** [comment]
:+1: looks good for now. It would be nice to move away from '`doc_type` checking.

**Comment by benrudolph** [comment]
@dannyroberts updated based on your comment, sorry that took absurdly long

**Pull request merged**